cms_coding_rules.md

```diff
@@ -88,10 +88,15 @@ If necessary to create a unique name, one can add the directory name:
 27. Provide meaningful argument names in method declarations in the header ﬁle to indicate usage, unless the type fully describes the usage.
 28. Try to avoid excessively long lines of code that impair readability.
 29. Data members of a class must not be redeﬁned in derived classes since doing so hides the original data member and could create confusion between the original and redefined data members.
+30. Use `std::abs()` instead of `fabs()`.
+31. Use C++ headers, e.g. `#include <cmath>` instead of `#include <math.h>`.
+32. Use MessageLogger instead of `std::cout` and `std::cerr` (\*). Allowed exception: standalone programs should use `std::cout` and `std::cerr`.
+33. Avoid `thread_local`.
 
 ## 5 -- Documentation Rules
 1. Always  comment  complex,  tricky,  or  non-intuitive  portions of  code.
 2. When revising code, be sure to update and revise comments.
+3. Do not leave commented-out code in files. Instead, use git versioning to retain old versions of code in the development history. If code is only commented out temporarily, a clear comment should be added describing why it is temporarily disabled and under what conditions it will be re-enabled.
```

Comment:
The "use git versioning to retain old versions of code" assumes that one knows to look in the git history for older versions of the code - people other than the original developer (who may very well have left CMS by the time the code is looked at again) may have no idea that an alternative version was ever available.
In addition, the very loose approach to a git history used in CMSSW makes it often very hard to figure out which commits contain useful code, and which ones are just iteration of non-working code before the final version is merged.

In addition, some code make a lot more sense being commented out than outright removed. For example, commented-out debug statements tell the people that read he code how to check some internal states, *etc.*
To those that advocate using `#ifdef`s for this, please explain how
```c++
#ifdef SOME_RANDOM_DEBUG_FLAG
    std::cout << "internal state: " << myObject.state() << std::endl;
#endif  // SOME_RANDOM_DEBUG_FLAG
```
is more readable or better maintainable than
```c++
    //std::cout << "internal state: " << myObject.state() << std::endl;
```
?

Reply:
It's more maintainable because the debug statement can be activated without editing the code. (I agree that the readability is not great.)

The key point is really in the last sentence of this code rule: files littered with commented-out code are hard to understand for someone looking at them for the first time, because it's usually not indicated *why* the commented-out code is there.

Reply:
I agree for the case where there are a bunch of statements that could be something somebody has tried and didn't work, or something that was there before and has been changed, etc.

However I don't think that
```c++
    //std::cout << "internal state: " << myObject.state() << std::endl;
```
is hard to understand - neither why the comment is there, nor what it would do if uncommented.

> It's more maintainable because the debug statement can be activated without editing the code.

True - unless there are `#define`/`#undef` for that symbol in the source file, then it does need to be edited.
Or unless you want to enable only one of those messages, not all of them.

Reply:
In my opinion 
```
//std::cout << "internal state: " << myObject.state() << std::endl;
```
is really hard to understand :-) This does not tell me if it was added during the development phase and user forgot to remove it or if it was really to indicate the it is for debugging purpose in future. Better to either use `#ifdef` or add a comment to indicate that why this comment is there.

Reply:
Then we disagree.

Reply:
Of course, for this specific example an ideal solution would be to have a *usable* system for debug statements, with a clear syntax, support for the most common idioms (not only the ones that the Core Software team deems "appropriate" for the developers to use), and a simply way of turning the debug information on and off.